src/crypto/pwhash/argon2_macros.rs

```diff
@@ -0,0 +1,197 @@
+pub fn gen_salt() -> Salt {
+    let mut salt = Salt([0; SALTBYTES]);
+    {
+        let Salt(ref mut sb) = salt;
+        randombytes_into(sb);
+    }
+    salt
+}
+
+/// The `derive_key()` function derives a key from a password and a `Salt`
+///
+/// The computed key is stored into key.
+///
+/// `opslimit` represents a maximum amount of computations to perform. Raising
+/// this number will make the function require more CPU cycles to compute a key.
+///
+/// `memlimit` is the maximum amount of RAM that the function will use, in
+/// bytes. It is highly recommended to allow the function to use at least 16
+/// megabytes.
+///
+/// For interactive, online operations, `OPSLIMIT_INTERACTIVE` and
+/// `MEMLIMIT_INTERACTIVE` provide a safe base line for these two
+/// parameters. However, using higher values may improve security.
+///
+/// For highly sensitive data, `OPSLIMIT_SENSITIVE` and `MEMLIMIT_SENSITIVE` can
+/// be used as an alternative. But with these parameters, deriving a key takes
+/// more than 10 seconds on a 2.8 Ghz Core i7 CPU and requires up to 1 gigabyte
+/// of dedicated RAM.
+///
+/// The salt should be unpredictable. `gen_salt()` is the easiest way to create a `Salt`.
+///
+/// Keep in mind that in order to produce the same key from the same password,
+/// the same salt, and the same values for opslimit and memlimit have to be
+/// used.
+///
+/// The function returns `Ok(key)` on success and `Err(())` if the computation didn't
+/// complete, usually because the operating system refused to allocate the
+/// amount of requested memory.
+pub fn derive_key<'a>(
+    key: &'a mut [u8],
+    passwd: &[u8],
+    &Salt(ref sb): &Salt,
+    OpsLimit(opslimit): OpsLimit,
+    MemLimit(memlimit): MemLimit,
+) -> Result<&'a [u8], ()> {
+    if unsafe {
+        $pwhash_name(
+            key.as_mut_ptr(),
+            key.len() as c_ulonglong,
+            passwd.as_ptr() as *const _,
+            passwd.len() as c_ulonglong,
+            sb as *const _,
+            opslimit as c_ulonglong,
+            memlimit,
+            VARIANT as c_int,
+        )
+    } == 0
+    {
+        Ok(key)
+    } else {
+        Err(())
+    }
+}
+
+
+/// The `pwhash()` returns a `HashedPassword` which
+/// includes:
+///
+/// - the result of a memory-hard, CPU-intensive hash function applied to the password
+///   `passwd`
+/// - the automatically generated salt used for the
+///   previous computation
+/// - the other parameters required to verify the password: opslimit and memlimit
+///
+/// `OPSLIMIT_INTERACTIVE` and `MEMLIMIT_INTERACTIVE` are safe baseline
+/// values to use for `opslimit` and `memlimit`.
+///
+/// The function returns `Ok(hashed_password)` on success and `Err(())` if it didn't complete
+/// successfully
+pub fn pwhash(
+    passwd: &[u8],
+    OpsLimit(opslimit): OpsLimit,
+    MemLimit(memlimit): MemLimit,
+) -> Result<HashedPassword, ()> {
+    let mut _out = HashedPassword([0; HASHEDPASSWORDBYTES]);
```

Comment:
why is it called `_out`?

Reply:
To suppress an unused mut warning.  I believe `out` needs to be mutable due to borrowing as a mutable pointer in the unsafe call. Will change this if not, or if there's a better way to do it

Reply:
Try `out.0.as_mut_ptr()`[:)](https://doc.rust-lang.org/std/primitive.slice.html#method.as_mut_ptr)